pop/service/service.go

```diff
@@ -224,47 +466,451 @@ func (s *Service) CheckConfig(req *network.Envelope) {
+	// to backward compatibility with orgs, who can't get hash of new final statement
+	for _, f := range finals {
+		// but signature on this conodes will be invalid
+		// because it's impossible to save signature on old hashes
+		s.data.Finals[string(f.Desc.Hash())] = final
+		// there is no need to support consistency of syncData and merge
+		// for old parties because their finalStatements are rewritten
+	}
+
+	s.save()
+	return true
+}
+
+/* --------------Propagation function-------------- */
+
+// PropagateFinal saves the new final statement
+func (s *Service) PropagateFinal(msg network.Message) {
+	fs, ok := msg.(*FinalStatement)
+	if !ok {
+		log.Error("Couldn't convert to a FinalStatement")
+		return
+	}
+	if err := fs.Verify(); err != nil {
+		log.Error(err)
+		return
+	}
+	*s.data.Finals[string(fs.Desc.Hash())] = *fs
+	s.save()
+	log.Lvlf2("%s Stored final statement %v", s.ServerIdentity(), fs)
+}
+
+/* ----------------Utilite functions--------------- */
+
+//signs FinalStatement with BFTCosi and Propagates signature to other nodes
+func (s *Service) signAndPropagate(final *FinalStatement, protoName string,
+	data []byte) onet.ClientError {
+	tree := final.Desc.Roster.GenerateNaryTreeWithRoot(2, s.ServerIdentity())
+	if tree == nil {
+		return onet.NewClientErrorCode(ErrorInternal,
+			"Root does not exist")
+	}
+	node, err := s.CreateProtocol(protoName, tree)
+	if err != nil {
+		return onet.NewClientError(err)
+	}
+
+	// Register the function generating the protocol instance
+	root, ok := node.(*bftcosi.ProtocolBFTCoSi)
+	if !ok {
+		return onet.NewClientErrorCode(ErrorInternal,
+			"protocol instance is invalid")
+	}
+
+	root.Msg, err = final.Hash()
+	if err != nil {
+		return onet.NewClientError(err)
+	}
+
+	root.Data = data
+	done := make(chan bool)
+	root.RegisterOnDone(func() {
+		done <- true
+	})
+	final.Signature = []byte{}
+	go node.Start()
+
+	select {
+	case <-done:
+		sig := root.Signature()
+		if len(sig.Sig) >= SIGSIZE {
+			final.Signature = sig.Sig[:SIGSIZE]
+		} else {
+			final.Signature = []byte{}
+		}
+	case <-time.After(timeout):
+		log.Error("signing failed on timeout")
+		return onet.NewClientErrorCode(ErrorTimeout,
+			"signing timeout")
+	}
+	if len(final.Signature) <= 0 {
+		log.Error("Signing failed")
+		return onet.NewClientErrorCode(ErrorTimeout,
+			"Signing failed")
+	}
+
+	replies, err := s.PropagateFinalize(final.Desc.Roster, final, 10000)
+	if err != nil {
+		return onet.NewClientError(err)
+	}
+	if replies != len(final.Desc.Roster.List) {
+		log.Warn("Did only get", replies)
+	}
+	s.save()
+	return nil
+}
+
+// Merge sends MergeConfig to all parties,
+// Receives Replies, updates info about global merge party
+// When all merge party's info is saved, merge it and starts global sighning process
+// After all, sends StoreConfig request to other conodes of own party
+func (s *Service) Merge(final *FinalStatement, m *merge) (*FinalStatement,
```

Comment:
I prefer the following structure of the file:
`init()`
constants and structure-definitions
public methods on structure
private methods on structure
package public functions
package private functions

Please re-arrange.